cloud-platform/app-deploy.md

```diff
@@ -0,0 +1,205 @@
+---
+category: cloud-platform
+expires: 2018-01-31
+---
+# Deploying an application to the Cloud Platform
+
+## Overview
+
+The aim of this guide is to walkthrough the process of deploying an application into the Cloud Platform.
+
+This guide uses a pre-configured application as an example of how to deploy your own.
+
+## Prerequisites
+
+This guide assumes the following:
+
+* Kubectl is installed and configured.
+* Docker is installed and configured.
+* Authentication with the cluster has been established.
+* An environment has been created within the Cloud Platform.
+* Access to AWS, with ECR upload permissions.
+* AWS CLI configured with account credentials.
+
+If you are not deploying your own application and would like to deploy the example application, clone the following repo:
+
+[https://github.com/ministryofjustice/cloud-platform-demo-app](https://github.com/ministryofjustice/cloud-platform-demo-app)
+
+## Pushing application to ECR
+
+To deploy an application to the Cloud Platform, firstly the application image needs to be retrievable from a repository.
+
+Amazon's ECR, within ECS is where all of the application images used by the Cloud Platform are stored.
+
+### Creating or choosing a repository
```

Comment:
Are the end-users supposed to have the necessary permissions to do this? If not, we could always manage this with terraform and PRs.

Reply:
One for @kalbir and @kerin I think.

Reply:
There's mention in [this doc](https://github.com/ministryofjustice/cloud-platform/blob/master/architecture-decision-record/007-Use-ECR-As-Container-Registry.md) on how user access can be implemented.

As far as I understand it's still WIP so I don't want to block this PR. I think it's fine to go as-is and we can revisit.

@kerin, thoughts?

Reply:
Good question - I think ultimately we want to allow users to interact with AWS directly in the console and on the command line, but that requires that we sort out AWS Federation and user credentials. Given that we're going to be handing S3, RDS etc. via Terraform it's probably simplest to do the same for ECR right now.

One thing to consider though - to stop this from getting blocked right now we could just create an IAM user with admin rights on ECR generally, and use those single set of creds across the board, and tackle per-image rights later - @kalbir @alkar ?

Reply:
Any thoughts on how will we share that cred with the user who is trying to upload their image to ECR?

Reply:
Should we leave it as-is for now and wait for proper auth federation? It would mean that there's no way to actually follow the guide, however, if we expect auth to be sorted in the very near future I think it wouldn't be a real blocker.

Even if we temporarily took on the task of issuing a new pair of IAM creds for every potential user until auth is complete, would that be a lot of overhead? I'd like to avoid going the shared credentials route if possible.

Reply:
Yep, let's go with the sharing a cred around right now. Next story off the backlog is for users to get access to AWS resources so we can make this bit a priority to work out. 

That means we will have to:

a) update this to say something along the lines of, "when you are uploading your image to ECR you'll need a credential from the cloud platform team, ask in the #cloud-platform-users channel in slack"
b) create a tightly scoped credential for this purposs -- allows a user to push to ECR and the resulting image is public so anyone can pull it. 
c) put the credential somewhere behind access control where we can share it

Agreed? @ollieanwyll are you happy to update the docs here with that?

Reply:
Having the Cloud Platform team create and issue creds for each user sounds good to me - we should probably do that via Terraform too so we keep track of what we've done in AWS.

Reply:
On reflection, I'd *strongly* prefer we did this–start like we mean to go on...

Reply:
Amended with 83b8b27